notebooks/03-query/11-joins.ipynb

```diff
@@ -0,0 +1,2475 @@
+      "2023-10-26 01:01:52.442454 <<<<<<<<<<<<< \n"
+     ]
+    },
+    {
+     "data": {
+      "text/plain": [
+       "datetime.datetime(2023, 10, 26, 1, 1, 52, 442454)"
+      ]
+     },
+     "execution_count": 20,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "sql='''\n",
+    "SELECT TIME_FLOOR(c.__time, 'P1D') as _date, \n",
+    "  c.age_group, \n",
+    "  count(distinct c.client_ip) as distint_ips\n",
+    "FROM clicks_enhanced c \n",
+    "WHERE c.__time > CURRENT_TIMESTAMP  - INTERVAL '1' DAY\n",
+    "GROUP BY 1, 2\n",
+    "ORDER BY 3 DESC\n",
+    "LIMIT 10\n",
+    "'''\n",
+    "\n",
+    "result = measure_query(sql, 20)\n",
+    "# saving all results for comparison \n",
+    "join_results.append({\"query_type\":\"simple_prejoin\",\"results\":result})\n",
+    "print_timestamp()\n",
+    "print(result)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "d013699b-c3b9-497d-9712-fdb536a93206",
+   "metadata": {},
+   "source": [
+    "...even with the subquery to calculate totals by age group by day, it still avoids two joins by using the pre-joined datasource:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 21,
+   "id": "97ef67de-20a9-4aa0-b835-a0142cd8a0bd",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Results = avg:352.1546 ms   min:327.998 ms  max:471.693 ms\n",
+      "2023-10-26 01:01:59.490510 <<<<<<<<<<<<< \n"
+     ]
+    },
+    {
+     "data": {
+      "text/plain": [
+       "datetime.datetime(2023, 10, 26, 1, 1, 59, 490510)"
+      ]
+     },
+     "execution_count": 21,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "sql = '''\n",
+    "SELECT TIME_FLOOR(c.__time, 'P1D') as _date, c.user_id,\n",
+    "  c.age_group, \n",
+    "  count(1) as user_clicks,\n",
+    "  count(1)*100.0/agg.clicks as age_group_contrib_pct\n",
+    "FROM clicks_enhanced c\n",
+    "     INNER JOIN\n",
+    "     (\n",
+    "         SELECT TIME_FLOOR(cc.__time, 'P1D') as _date,\n",
+    "                cc.age_group,\n",
+    "                COUNT(1) as clicks\n",
+    "           FROM clicks_enhanced cc\n",
+    "           GROUP BY 1, 2\n",
+    "     ) agg\n",
+    "        ON agg._date = TIME_FLOOR(c.__time, 'P1D') AND agg.age_group=c.age_group \n",
+    "WHERE c.__time > CURRENT_TIMESTAMP  - INTERVAL '1' DAY\n",
+    "GROUP BY 1, 2, 3, agg.clicks \n",
+    "ORDER BY 1, 5 DESC\n",
+    "LIMIT 10\n",
+    "'''\n",
+    "result = measure_query(sql, 20)\n",
+    "# saving all results for comparison \n",
+    "join_results.append({\"query_type\":\"subquery_prejoin\",\"results\":result})\n",
+    "print_timestamp()\n",
+    "print(result)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "64372aea-9915-4e8a-a4ec-b34b160bd3b3",
+   "metadata": {},
+   "source": [
+    "### Results by Join Strategy"
```

Comment:
I like the performance stuff in this notebook. Suggest that you pull this, together with the function above, into a concluding section at the bottom like, "Assessing JOIN performance" or something - to avoid it becoming distracting from the main purpose of showing someone how to JOIN data - rather than as an interactive piece of research.

Reply:
The main purpose is not showing "how" to join data, that's here:  https://druid.apache.org/docs/latest/querying/datasource#join
The main purpose is to learn about how the join strategy affects processing and performance.